### ISSUE TITLE
[feature request] : 3 new great tts models

### ISSUE THREAD

**Issue by marwanlhabti5-coder** [CrispStrobe/CrispASR#377]
Please review and include those great open tts model : 

https://huggingface.co/netease-youdao/Confucius4-TTS

https://huggingface.co/KRAFTON/Raon-OpenTTS-1B

https://huggingface.co/FireRedTeam/FireRedTTS3 (smaller quantized : [int8](https://huggingface.co/drbaph/FireRedTTS3-int8/tree/main) )

**Comment by CrispStrobe** [comment]
Confucius4-TTS (#377): Architecture:
- T2S: GPT-2 (24L/1280d/20h) with TextEmbeddingProjector + ECAPA-TDNN speaker encoder + learned positional embeddings. Generates semantic codes (vocab 8194) from text + speaker conditioning.
- S2A: Flow-matching DiT (13-depth) + WaveNet (8-layer) → 80-band mel → BigVGAN vocoder @ 22050 Hz.
- External: Wav2Vec2-BERT 2.0 (layer 17 features, z-normalised) + CAMPPlus (192-dim speaker embedding).

Port will be attempted.

**Raon-OpenTTS-1B** — CC BY-NC 4.0, skipping.

**FireRedTTS3** — Apache 2.0 but ~21 GB, 4+ component pipeline. Deferred atm.

**Comment by subof** [comment]
Did you manage to add the confucius4-tts engine? 
I tested it, and it turned out to be one of the best.

**Comment by CrispStrobe** [comment]
Yes — Confucius4-TTS is now implemented and, importantly, model-backed rather than just compile-tested.

The strongest current evidence is:

- fully native zero-shot pipeline: **8/8 ASR round-trip**, with no Python or environment escape paths;
- separate CUDA P100 live run: **8/8 verbatim**, 183.2 s versus 685.2 s CPU (**3.74x**);
- bounded CPU live run: verbatim round-trip at **2.23 GB peak RSS**;
- conditioned S2A parity: cosine **1.000000 at every stage** across the 947-frame prompt path;
- Metal model-backed validation also transcribes verbatim.

The reproducible GPU proof is checked in at `tools/kaggle/confucius4-gpu-verify/`; the detailed run ledger is `docs/confucius4-tts/PLAN.md`. Usage is:

```bash
crispasr --backend confucius4-tts -m auto \
  --tts "Hello from Confucius." -l en \
  --voice ref.wav --i-have-rights --tts-output out.wav
```

I also corrected stale `docs/tts.md` wording on `main`: w2v-BERT conditioning is native now and `-m auto` downloads its companion. So the answer for Confucius is an evidence-backed **yes**. Raon remains intentionally skipped because its weights are CC-BY-NC-4.0; FireRedTTS3 remains the unimplemented/deferred part of this multi-model request because of its roughly 21 GB, four-component pipeline.

**Comment by subof** [comment]
Thank you very much. I also tested FireRedTTS3 and found it to be on par with Qwen3-TTS, with a strong accent due to cross-lingual transfere. Confucius4-TTS does a much better job.

**Comment by subof** [comment]
@niksedk , Could you add Confucius4-TTS to your program, please? 

Thank you.

**Comment by subof** [comment]
The only engine better than this one is Higgs-Audio-v3-4b, which you abandoned due to licensing issues. Higgs is designed for film dubbing and supports emotion tags. It also produces high-quality audio. But it's used by other programs for voice-over and subtitling.

**Comment by niksedk** [comment]
>@niksedk , Could you add Confucius4-TTS to your program, please?

Tried Confucius4-TTS but it was very robotic for me... might have some bugs?

**Comment by subof** [comment]
> > [@niksedk](https://github.com/niksedk) , Could you add Confucius4-TTS to your program, please?
> 
> Tried Confucius4-TTS but it was very robotic for me... might have some bugs?

The robotic sound usually happens for two reasons, both easily fixable:

1. Missing Reference Audio: Unlike other engines, [Confucius4](https://www.google.com/url?sa=i&source=web&rct=j&url=https://arxiv.org/html/2608.11650v1&ved=2ahUKEwjd0c2K0s2WAxX83AIHHdnnCHoQ1fkOegkIAggBCN0BECs&opi=89978449&cd&psig=AOvVaw1iljiEp836ffKdS53EO3ze&ust=1788361017313000) requires a high-quality 3-second reference audio clip (Zero-shot) to build the prosody. Without it (or with a low-res default), it defaults to a flat, robotic synthetic tone.
2. Quantization Drop: If you tested the 4-bit (q4) version, the flow-matching alignment breaks down heavily. It sounds pristine and natural on FP16 / INT8, perfectly holding accents and intonations.

It’s definitely worth implementing! It’s currently the most stable open-source engine for clear, non-dramatic narrative reading, especially for non-English voiceovers and for cross‑language transfer.

The model weighs less than the flagship versions of MOSS (8B) and Higgs (4B), but thanks to its integration with the vLLM backend (PagedAttention architecture), it achieves the lowest word error rate (WER/CER) on the Seed-TTS and MiniMax benchmarks. The synthesis quality and speech smoothness are on par with commercial heavyweights.

**Comment by CrispStrobe** [comment]
FireRedTTS3 is on `main` — the last of the three models in this issue
(Confucius4-TTS and Raon-OpenTTS shipped earlier).

`--backend fireredtts3`, Apache-2.0, zero-shot voice cloning: Qwen3 backbone over
continuous speech representations + the RedAE acoustic autoencoder + a CAM++
speaker encoder. F16 and Q4_K GGUFs on `cstr/fireredtts3-GGUF`.

**Per-stage parity against the reference**, all gates met:

| stage | cos |
|---|---|
| `penc_prompt`, `prefill_embeds`, `llm_prefill_out`, `stop_scores`, `latents_gen`, `dec_hidden` | 1.000000 |
| `gen_audio` | 0.999999 |
| `spk_llm` | 1.000000 |
| `spk_dit` | 0.999991 |
| `spk_emb` | 0.999452 |
| `campp_fbank` | 0.997589 |

**The interesting part was the speaker encoder, and the bug was not in this
backend at all.** `spk_emb` started at cos **0.268** with magnitude 1.67× the
reference. The decisive measurement was feeding the stage the *reference* fbank:
`spk_emb_orafb` still failed at 0.278 while the fbank itself read 0.9976. Bad
output from known-good input indicts the network, not its input.

The cause was in shared CAM++ code (`src/chatterbox_campplus.cpp`). `seg_pooling`
is `F.avg_pool1d(k=100, stride=100, ceil_mode=True)`, and the port divided
*every* window by 100 — including the trailing partial one. Torch divides the
tail by the frames actually in it. Settled by running torch rather than by
reading `count_include_pad`, whose flag semantics the old comment had reasoned
its way to the wrong answer on:

```
F.avg_pool1d(torch.ones(1,1,551), kernel_size=100, stride=100, ceil_mode=True)
  -> all six segments are exactly 1.0        (the wrong rule gives 0.51)
```

Fixing it moved `spk_emb_orafb` from 0.277717 to **0.999997** and `spk_emb` from
0.268065 to **0.999452** — it was the entire divergence.

That code is shared by **chatterbox, confucius4, cosyvoice3, dots and
fireredtts3**, all of which shipped with it. It survived because every one of
those was accepted end-to-end (ASR roundtrip), and the error only touched the
trailing window — at T=551 a 0.51× context over 9% of frames, compounded through
52 dense layers into a sigmoid gate. A roundtrip cannot localise, so it cannot
see a defect a later stage partially absorbs. It is now pinned by
`tests/test-campplus-segpool.cpp`, which asserts the all-ones tail is 1.0 and not
0.51, with a ramp control so that case cannot pass for a trivial reason.

**On the remaining `campp_fbank` 0.997589 — it is the resampler, not a fbank
convention error, and it should not be chased as one.** Both candidates were
measured rather than argued: unscaled vs int16-scaled input costs cos 0.952 / 11%
magnitude, an order too large, and the reference feeds torchaudio `[-1,1]`
floats, so `int16_scale=false` is correct. Running one clip through
`torchaudio.functional.resample` vs polyphase and then through the *identical*
kaldi fbank gives cos 0.998250 / 0.4% — the same order as observed. Its cost is
bounded by measurement, not assumption: `spk_emb` scores 0.999452 on our fbank
against 0.999997 on the oracle, so the resampler delta is worth 0.0005 of
embedding cosine. Matching it exactly would mean reimplementing torchaudio's
Kaiser-windowed sinc, which is not worth that. The gate is set at 0.995 with the
reasoning recorded in the source; a real scaling error lands near 0.95 and a
layout error near 0, so the gate still separates what it must.

**Two fixes CI surfaced after the merge**, both real and neither a merge
artifact — the branch's own Windows run had died at the build step, so this was
the first time the full matrix ran against the port:

- Under `GGML_BACKEND_DL` the backend failed to link (`undefined reference to
  ggml_backend_is_cpu` / `ggml_backend_cpu_set_n_threads`). That config makes
  the CPU backend a dlopen'd module so those symbols are not linkable;
  `src/core/ggml_cpu_backend.h` is the shim for it, and the file already
  included it — only the `set_n_threads` calls were raw.
- `tests/test-arch-backend-map.cpp` ("every emitted backend name is a name a
  surface can open") failed on `fireredtts3` and `fireredtts3-redae`: entries
  were added to `src/core/arch_backend_map.h` without adding the backend to the
  test's `known` list.

`docs/architecture.md` now has a `### fireredtts3` section — `docs/tts.md`
linked to it and it did not exist, which the doc-anchor check caught. It records
the CAM++ finding and the fbank gate reasoning so neither gets re-derived.

With that, this issue's three models are all shipped: Confucius4-TTS,
Raon-OpenTTS and now FireRedTTS3.

**Comment by CrispStrobe** [comment]
All three have shipped.

| model | backend | first release |
|---|---|---|
| [Confucius4-TTS](https://huggingface.co/netease-youdao/Confucius4-TTS) | `--backend confucius4-tts` | **v0.8.30** |
| [Raon-OpenTTS](https://huggingface.co/KRAFTON/Raon-OpenTTS-1B) | `--backend raon` (0.3B) / `--backend raon-1b` (1B) | **v0.8.32** |
| [FireRedTTS3](https://huggingface.co/FireRedTeam/FireRedTTS3) | `--backend fireredtts3` | **v0.8.33** |

A few notes that matter in practice:

- **Confucius4-TTS is zero-shot only.** Without `--voice` conditioning the output is unintelligible *by design* — that is the model, not a port defect. Use `--voice ref.wav --i-have-rights`. 14 languages via `-l <lang>`.
- **Raon shipped in both sizes.** The 1B needed a real fix first: its DiT attention accumulated the KQ product in F16 through ggml's fused flash kernel, and `ggml_flash_attn_ext_set_prec(F32)` is *silently ignored* on P100/sm_60 — so the 1B drifted to NaN on that hardware while the 0.3B survived. Manual F32 SDPA is now the default for f5-family DiT attention and flash is opt-in (`CRISPASR_F5_FLASH=1`), so this is correct on every backend rather than on an allowlist of cards.
- **Both Raon sizes are CC-BY-NC-4.0**, non-commercial only, and the registry prints that on first download.
- FireRedTTS3 is zero-shot cloning over continuous 64-d 25 Hz RedAE latents with a Qwen3 backbone, Apache-2.0.

Registry URLs for all of these were just re-checked with `tools/check-registry-urls.py` and resolve.

Closing. If any of them misbehaves in practice, a new issue with the command and output is more useful than reopening this one, since the three have nothing in common at runtime.
